File: sources/dart/10_pi/main.dart
// task 10 pi — expected output: 44889
// build: dart compile exe -o prog main.dart (aot; jit has no build step)    run: ./prog (aot) | dart main.dart (jit)
// Gibbons' unbounded spigot on hand-written base-10^9 limbs: dart:core has no big integers.
// Every intermediate stays below 2^53, so the arithmetic is exact if compiled to JavaScript too.

const int _base = 1000000000;

/// Signed arbitrary-precision integer: little-endian base-10^9 magnitude limbs.
/// The magnitude carries no trailing zero limbs; an empty magnitude is zero.
class Big {
  final List<int> m;
  final bool neg;

  const Big._(this.m, this.neg);

  static const Big zero = Big._(<int>[], false);

  static Big of(List<int> mag, bool negative) {
    int n = mag.length;
    while (n > 0 && mag[n - 1] == 0) {
      n--;
    }
    if (n == 0) {
      return zero;
    }
    return Big._(n == mag.length ? mag : mag.sublist(0, n), negative);
  }

  static Big fromInt(int v) {
    if (v == 0) {
      return zero;
    }
    return Big.of(<int>[v < 0 ? -v : v], v < 0);
  }

  static int cmpMag(List<int> a, List<int> b) {
    if (a.length != b.length) {
      return a.length < b.length ? -1 : 1;
    }
    for (int i = a.length - 1; i >= 0; i--) {
      if (a[i] != b[i]) {
        return a[i] < b[i] ? -1 : 1;
      }
    }
    return 0;
  }

  static int cmp(Big a, Big b) {
    if (a.neg != b.neg) {
      return a.neg ? -1 : 1;
    }
    final int c = cmpMag(a.m, b.m);
    return a.neg ? -c : c;
  }

  static List<int> addMag(List<int> a, List<int> b) {
    final int n = a.length > b.length ? a.length : b.length;
    final List<int> r = List<int>.filled(n + 1, 0);
    int carry = 0;
    for (int i = 0; i < n; i++) {
      int v = carry;
      if (i < a.length) {
        v += a[i];
      }
      if (i < b.length) {
        v += b[i];
      }
      if (v >= _base) {
        v -= _base;
        carry = 1;
      } else {
        carry = 0;
      }
      r[i] = v;
    }
    if (carry != 0) {
      r[n] = carry;
      return r;
    }
    return r.sublist(0, n);
  }

  /// a - b for a >= b, both non-negative.
  static List<int> subMag(List<int> a, List<int> b) {
    final int n = a.length;
    final List<int> r = List<int>.filled(n, 0);
    int borrow = 0;
    for (int i = 0; i < n; i++) {
      int v = a[i] - borrow - (i < b.length ? b[i] : 0);
      if (v < 0) {
        v += _base;
        borrow = 1;
      } else {
        borrow = 0;
      }
      r[i] = v;
    }
    return r;
  }

  static Big add(Big a, Big b) {
    if (a.neg == b.neg) {
      return Big.of(addMag(a.m, b.m), a.neg);
    }
    final int c = cmpMag(a.m, b.m);
    if (c == 0) {
      return zero;
    }
    if (c > 0) {
      return Big.of(subMag(a.m, b.m), a.neg);
    }
    return Big.of(subMag(b.m, a.m), b.neg);
  }

  static Big negate(Big a) => a.m.isEmpty ? a : Big._(a.m, !a.neg);

  static Big sub(Big a, Big b) => add(a, negate(b));

  /// Multiply by a small non-negative integer.
  static Big mulSmall(Big a, int m) {
    if (m == 0 || a.m.isEmpty) {
      return zero;
    }
    final List<int> r = List<int>.filled(a.m.length + 1, 0);
    int carry = 0;
    for (int i = 0; i < a.m.length; i++) {
      final int v = a.m[i] * m + carry;
      r[i] = v % _base;
      carry = v ~/ _base;
    }
    int n = a.m.length;
    if (carry != 0) {
      r[n] = carry;
      n++;
    }
    return Big.of(r.sublist(0, n), a.neg);
  }

  /// floor(x / y) for y > 0 and a small quotient, by repeated subtraction.
  static int divSmall(Big x, Big y) {
    int q = 0;
    if (x.neg) {
      while (cmp(x, zero) < 0) {
        x = add(x, y);
        q -= 1;
      }
    } else {
      final Big y10 = mulSmall(y, 10);
      while (cmp(x, y10) >= 0) {
        x = sub(x, y10);
        q += 10;
      }
      while (cmp(x, y) >= 0) {
        x = sub(x, y);
        q += 1;
      }
    }
    return q;
  }
}

void main() {
  Big q = Big.fromInt(1);
  Big r = Big.fromInt(0);
  Big t = Big.fromInt(1);
  int k = 1;
  int n = 3;
  int l = 3;

  const int digits = 10000;
  int emitted = 0;
  int digitSum = 0;
  while (emitted < digits) {
    final Big nt = Big.mulSmall(t, n);
    final Big lhs = Big.sub(Big.add(Big.mulSmall(q, 4), r), t);
    if (Big.cmp(lhs, nt) < 0) {
      digitSum += n;
      emitted++;
      final Big q2 = Big.mulSmall(q, 10);
      final Big r2 = Big.mulSmall(Big.sub(r, nt), 10);
      n = Big.divSmall(Big.mulSmall(Big.add(Big.mulSmall(q, 3), r), 10), t) - 10 * n;
      q = q2;
      r = r2;
    } else {
      final Big q2 = Big.mulSmall(q, k);
      final Big r2 = Big.mulSmall(Big.add(Big.mulSmall(q, 2), r), l);
      final Big t2 = Big.mulSmall(t, l);
      n = Big.divSmall(Big.add(Big.mulSmall(q, 7 * k + 2), Big.mulSmall(r, l)), t2);
      q = q2;
      r = r2;
      t = t2;
      k += 1;
      l += 2;
    }
  }
  print(digitSum);
}

File: sources/kotlin/10_pi/main.kt
// task 10 pi — expected output: 44889
// build: kotlinc main.kt -include-runtime -d prog.jar    run: java -jar prog.jar    [native build: kotlinc-native -opt -o prog main.kt    native run: ./prog]
// note: the common Kotlin stdlib has no big integers (java.math is JVM-only), so this is a hand-written base-10^9 limb bignum: multiply by a small int, add/subtract, and divide by a big divisor whose quotient is a small int. Gibbons' unbounded spigot.

private const val BASE = 1000000000L

private fun magCmp(a: LongArray, al: Int, b: LongArray, bl: Int): Int {
    if (al != bl) return if (al > bl) 1 else -1
    var i = al - 1
    while (i >= 0) {
        if (a[i] != b[i]) return if (a[i] > b[i]) 1 else -1
        i--
    }
    return 0
}

// Signed arbitrary-precision integer, little-endian base-10^9 limbs in mag[0 until len].
private class Big {
    var neg = false
    var mag = LongArray(4)
    var len = 0

    private fun grow(n: Int) {
        var cap = mag.size
        while (cap < n) cap *= 2
        val a = LongArray(cap)
        mag.copyInto(a, 0, 0, len)
        mag = a
    }

    private fun trim() {
        while (len > 0 && mag[len - 1] == 0L) len--
        if (len == 0) neg = false
    }

    fun setLong(v: Long) {
        var x = v
        neg = x < 0L
        if (neg) x = -x
        var i = 0
        while (x > 0L) {
            if (i == mag.size) grow(i + 1)
            mag[i] = x % BASE
            x /= BASE
            i++
        }
        len = i
        if (len == 0) neg = false
    }

    fun set(o: Big) {
        if (mag.size < o.len) grow(o.len)
        o.mag.copyInto(mag, 0, 0, o.len)
        len = o.len
        neg = o.neg
    }

    fun setMulSmall(o: Big, m: Long) {
        if (o.len == 0 || m == 0L) {
            len = 0
            neg = false
            return
        }
        var mm = m
        neg = o.neg
        if (mm < 0L) {
            neg = !neg
            mm = -mm
        }
        if (mag.size < o.len + 1) grow(o.len + 1)
        var carry = 0L
        var i = 0
        while (i < o.len) {
            val p = o.mag[i] * mm + carry
            mag[i] = p % BASE
            carry = p / BASE
            i++
        }
        len = o.len
        while (carry > 0L) {
            if (len == mag.size) grow(len + 1)
            mag[len] = carry % BASE
            carry /= BASE
            len++
        }
    }

    fun mulSmall(m: Long) {
        if (len == 0 || m == 0L) {
            len = 0
            neg = false
            return
        }
        var mm = m
        if (mm < 0L) {
            neg = !neg
            mm = -mm
        }
        if (mm == 1L) return
        var carry = 0L
        var i = 0
        while (i < len) {
            val p = mag[i] * mm + carry
            mag[i] = p % BASE
            carry = p / BASE
            i++
        }
        while (carry > 0L) {
            if (len == mag.size) grow(len + 1)
            mag[len] = carry % BASE
            carry /= BASE
            len++
        }
    }

    fun cmp(o: Big): Int {
        if (neg != o.neg) return if (neg) -1 else 1
        val c = magCmp(mag, len, o.mag, o.len)
        return if (neg) -c else c
    }

    // this += |o|, sign unchanged
    private fun addMag(o: Big) {
        var carry = 0L
        val n = if (len > o.len) len else o.len
        if (mag.size < n + 1) grow(n + 1)
        var i = 0
        while (i < n) {
            val a = if (i < len) mag[i] else 0L
            val b = if (i < o.len) o.mag[i] else 0L
            var s = a + b + carry
            if (s >= BASE) {
                s -= BASE
                carry = 1L
            } else {
                carry = 0L
            }
            mag[i] = s
            i++
        }
        len = n
        if (carry != 0L) {
            mag[len] = carry
            len++
        }
    }

    // this -= o, requires |this| > |o|; sign unchanged
    private fun subMagLow(o: Big) {
        var borrow = 0L
        var i = 0
        while (i < len) {
            val b = if (i < o.len) o.mag[i] else 0L
            var d = mag[i] - b - borrow
            if (d < 0L) {
                d += BASE
                borrow = 1L
            } else {
                borrow = 0L
            }
            mag[i] = d
            i++
        }
        trim()
    }

    // this := |o| - |this|, requires |o| > |this|
    private fun subMagHigh(o: Big) {
        if (mag.size < o.len) grow(o.len)
        var borrow = 0L
        var i = 0
        while (i < o.len) {
            val a = if (i < len) mag[i] else 0L
            var d = o.mag[i] - a - borrow
            if (d < 0L) {
                d += BASE
                borrow = 1L
            } else {
                borrow = 0L
            }
            mag[i] = d
            i++
        }
        len = o.len
        trim()
    }

    fun addAssign(o: Big) {
        if (o.len == 0) return
        if (len == 0) {
            set(o)
            return
        }
        if (neg == o.neg) {
            addMag(o)
            return
        }
        val c = magCmp(mag, len, o.mag, o.len)
        if (c == 0) {
            len = 0
            neg = false
        } else if (c > 0) {
            subMagLow(o)
        } else {
            subMagHigh(o)
            neg = o.neg
        }
    }

    fun subAssign(o: Big) {
        if (o.len == 0) return
        if (len == 0) {
            set(o)
            neg = !neg
            return
        }
        if (neg != o.neg) {
            addMag(o)
            return
        }
        val c = magCmp(mag, len, o.mag, o.len)
        if (c == 0) {
            len = 0
            neg = false
        } else if (c > 0) {
            subMagLow(o)
        } else {
            subMagHigh(o)
            neg = !neg
        }
    }
}

// floor(num / den) as a Long, with den > 0. The quotient is a small integer here, so a
// doubling search plus a binary search is enough and stays exact.
private fun divFloor(num: Big, den: Big, prod: Big): Long {
    if (num.len == 0) return 0L
    val negNum = num.neg
    if (negNum) num.neg = false
    var q = 0L
    if (num.cmp(den) >= 0) {
        var hi = 1L
        prod.setMulSmall(den, hi)
        while (num.cmp(prod) >= 0) {
            hi *= 2L
            prod.setMulSmall(den, hi)
        }
        var lo = hi / 2L
        while (lo + 1L < hi) {
            val mid = (lo + hi) / 2L
            prod.setMulSmall(den, mid)
            if (num.cmp(prod) >= 0) lo = mid else hi = mid
        }
        q = lo
    }
    if (negNum) {
        prod.setMulSmall(den, q)
        if (num.cmp(prod) != 0) q += 1L
        q = -q
        num.neg = true
    }
    return q
}

fun main() {
    // Gibbons' unbounded spigot: (q, r, t, k, n, l), n is the digit produced when safe.
    val q = Big()
    q.setLong(1L)
    val r = Big()
    r.setLong(0L)
    val t = Big()
    t.setLong(1L)
    var k = 1L
    var n = 3L
    var l = 3L

    val lhs = Big()
    val rhs = Big()
    val num = Big()
    val den = Big()
    val prod = Big()
    val tmp = Big()

    var digits = 0
    var sum = 0L
    while (digits < 10000) {
        // 4*q + r - t < n*t  means the digit n is safe to emit
        lhs.setMulSmall(q, 4L)
        lhs.addAssign(r)
        lhs.subAssign(t)
        rhs.setMulSmall(t, n)
        if (lhs.cmp(rhs) < 0) {
            sum += n
            digits++
            // n = div(10*(3*q + r), t) - 10*n ; q = 10*q ; r = 10*(r - n*t)
            num.setMulSmall(q, 3L)
            num.addAssign(r)
            num.mulSmall(10L)
            val next = divFloor(num, t, prod) - 10L * n
            tmp.setMulSmall(t, n)
            r.subAssign(tmp)
            r.mulSmall(10L)
            q.mulSmall(10L)
            n = next
        } else {
            // q = q*k ; r = (2*q + r)*l ; t = t*l ; n = div(q*(7*k + 2) + r*l, t*l)
            den.setMulSmall(t, l)
            num.setMulSmall(r, l)
            tmp.setMulSmall(q, 7L * k + 2L)
            num.addAssign(tmp)
            n = divFloor(num, den, prod)
            tmp.setMulSmall(q, 2L)
            tmp.addAssign(r)
            tmp.mulSmall(l)
            r.set(tmp)
            q.mulSmall(k)
            t.set(den)
            k += 1L
            l += 2L
        }
    }
    println(sum)
}
